src/mlpack/methods/ann/gan/gan.hpp

```diff
@@ -0,0 +1,357 @@
+/**
+ * @file gan.hpp
+ * @author Kris Singh
+ * @author Shikhar Jaiswal
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license. You should have received a copy of the
+ * 3-clause BSD license along with mlpack. If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_GAN_GAN_HPP
+#define MLPACK_METHODS_ANN_GAN_GAN_HPP
+
+#include <mlpack/core.hpp>
+
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/gan/gan_policies.hpp>
+#include <mlpack/methods/ann/visitor/output_parameter_visitor.hpp>
+#include <mlpack/methods/ann/visitor/reset_visitor.hpp>
+#include <mlpack/methods/ann/visitor/weight_size_visitor.hpp>
+#include <mlpack/methods/ann/visitor/weight_set_visitor.hpp>
+
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. **/ {
+
+/**
+ * The implementation of the standard GAN module. Generative Adversarial
+ * Networks (GANs) are a class of artificial intelligence algorithms used
+ * in unsupervised machine learning, implemented by a system of two neural
+ * networks contesting with each other in a zero-sum game framework. This
+ * technique can generate photographs that look at least superficially
+ * authentic to human observers, having many realistic characteristics.
+ *
+ * For more information, see the following paper:
+ *
+ * @code
+ * @article{Goodfellow14,
+ *   author    = {Ian J. Goodfellow, Jean Pouget-Abadi, Mehdi Mirza, Bing Xu,
+ *                David Warde-Farley, Sherjil Ozair, Aaron Courville and
+ *                Yoshua Bengio},
+ *   title     = {Generative Adversarial Nets},
+ *   year      = {2014},
+ *   url       = {http://arxiv.org/abs/1406.2661},
+ *   eprint    = {1406.2661},
+ * }
+ * @endcode
+ *
+ * @tparam Model The class type of Generator and Discriminator.
+ * @tparam InitializationRuleType Type of Initializer.
+ * @tparam Noise The noise function to use.
+ * @tparam PolicyType The GAN variant to be used (GAN, DCGAN, WGAN or WGANGP).
+ */
+template<
+  typename Model,
+  typename InitializationRuleType,
+  typename Noise,
+  typename PolicyType = StandardGAN
+>
+class GAN
+{
+ public:
+  /**
+   * Constructor for GAN class.
+   *
+   * @param trainData The real data.
+   * @param generator Generator network.
+   * @param discriminator Discriminator network.
+   * @param batchSize Batch size to be used for training.
+   * @param generatorUpdateStep Number of steps to train Discriminator
+   *                            before updating Generator.
+   * @param preTrainSize Number of pre-training steps of Discriminator.
+   * @param multiplier Ratio of learning rate of Discriminator to the Generator.
+   * @param clippingParameter Weight range for enforcing Lipschitz constraint.
+   * @param lambda Parameter for setting the gradient penalty.
+   */
+  GAN(arma::mat& trainData,
```

Comment:
Should this be `const`?

Reply:
We can make it to be `const`, however I had checked through all the layers, and wherever the data is passed, no changes would be done on it. I'll make the changes.